src/Plugin/KeyProvider/PubkeyEncryptKeyProvider.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\pubkey_encrypt\Plugin\KeyProvider\PubkeyEncryptKeyProvider.
+ */
+
+namespace Drupal\pubkey_encrypt\Plugin\KeyProvider;
+
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\key\Plugin\KeyProviderBase;
+use Drupal\key\Plugin\KeyPluginFormInterface;
+use Drupal\key\Plugin\KeyProviderSettableValueInterface;
+use Drupal\key\Exception\KeyValueNotSetException;
+use Drupal\key\KeyInterface;
+use Drupal\user\Entity\Role;
+use Drupal\Core\Session\AccountInterface;
+
+/**
+ * Adds a key provider as per the requirements of Pubkey Encrypt module.
+ *
+ * @KeyProvider(
+ *   id = "pubkey_encrypt",
+ *   label = @Translation("Pubkey Encrypt"),
+ *   description = @Translation("Stores and Retrieves the key as per the requirements of Pubkey Encrypt module."),
+ *   storage_method = "pubkey_encrypt",
+ *   key_value = {
+ *     "accepted" = TRUE,
+ *     "required" = FALSE
+ *   }
+ * )
+ */
+class PubkeyEncryptKeyProvider extends KeyProviderBase implements KeyPluginFormInterface, KeyProviderSettableValueInterface {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $roleOptions = [];
+    foreach (Role::loadMultiple() as $role) {
+      $roleOptions[$role->id()] = $role->label();
+    }
+    unset($roleOptions[AccountInterface::ANONYMOUS_ROLE]);
+    unset($roleOptions[AccountInterface::AUTHENTICATED_ROLE]);
+
+    $form['role'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Role'),
+      '#description' => $this->t('Share keys would be generated and stored for all the users in this Role.'),
+      '#options' => $roleOptions,
+      '#default_value' => $this->getConfiguration()['role'],
+      '#required' => TRUE,
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateConfigurationForm(array &$form, FormStateInterface $form_state) {
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state) {
+    $this->setConfiguration($form_state->getValues());
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getKeyValue(KeyInterface $key) {
+    $key_value = '';
+
+    $currentUserId = \Drupal::currentUser()->id();
+
+    // Retrieve the actual key value from the Share key of user.
+    $shareKeys = $this->configuration['share_keys'];
+    if (isset($shareKeys[$currentUserId])) {
+      $shareKey = base64_decode($shareKeys[$currentUserId]);
```

Comment:
Base64 encoding/decoding a stored is completely unnecessary here. I did it only because then viewing the key values from database/configuration system looked a bit nice if the keys are base64_encoded. We should remove it if it is affecting performance. What do you say?